Cubical/Data/Fin/Base.agda

```diff
@@ -0,0 +1,57 @@
+{-# OPTIONS --cubical --safe #-}
+
+module Cubical.Data.Fin.Base where
+
+open import Cubical.Core.Everything
+
+open import Cubical.Foundations.Function
+open import Cubical.Foundations.HLevels
+
+open import Cubical.Data.Empty
+open import Cubical.Data.Nat
+open import Cubical.Data.Nat.Order
+open import Cubical.Data.Sum
+
+open import Cubical.Relation.Nullary
+
+Fin : ℕ → Set
+Fin n = Σ[ k ∈ ℕ ] k < n
```

Comment:
Is there some particular reason you are using this definition? There is the other inductive family definition that I usually find behaves better.

Reply:
Namely this definition:
```
data Fin : (n : N) -> Set where
   zero : (n : N) -> Fin (suc n)
   suc : (n : N) -> Fin n -> Fin (suc n)
```

Reply:
@riaqn that definitionis an inductive family, which Agda doesn't support fully in cubical mode yet. For a different definition of Fin that would work you could define it recursively as a coproduct of unit types. I don't know the trade offs between that definition and the definition used here though.

Reply:
@potato4444 Hmm, I didn't know that. The definition itself passes the type check. Do you have some example results of this Fin that couldn't be accomplished in the current cubical agda?

Reply:
@riaqn this is what doesn't work https://github.com/agda/cubical/pull/57#issuecomment-461174095

Reply:
Yeah, the lack of support is why I avoided it. I think there may be other things that don't work, although possibly they have been fixed since I encountered them. For instance, a while ago I was playing around with the usual `Fin` definition, and cubical agda wouldn't infer the value of implicit arguments `k` from explicit `Fin k` arguments.

This is also why I chose the `k + m = n` definition of natural ordering instead of the usual indexed  inductive definition.

Reply:
> @riaqn that definition is an inductive family, which Agda doesn't support fully in cubical mode yet. For a different definition of Fin that would work you could define it recursively as a coproduct of unit types. I don't know the trade offs between that definition and the definition used here though.

But I assume this issue should only be temporary? If I understand it correctly, it only needs the computation rule for the least fixpoint, since an inductive family is the initial algebra for a dependent polynomial functor/indexed container.

Reply:
>For instance, a while ago I was playing around with the usual Fin definition, and cubical agda wouldn't infer the value of implicit arguments k from explicit Fin k arguments.

@dolio That's weird, if it happens again please report it. It could just be that the unification constraint wasn't in the pattern fragment (maybe because it only held for some (i = i0)), but otherwise type constructors should behave the same wrt type inference.

@L-TChen The problem is constructors that specialize the index argument of their result type, they need special care for `transp`.

Reply:
As @Saizan and others said already: inductive families are currently not supported by Cubical Agda in the sense that `transp` does not reduce correctly (or at all?) for them. There is no deep theoretical reason for this; it's "only" a matter of extending the internal definition of `transp` in Agda following http://www.cs.cmu.edu/~ecavallo/works/popl19.pdf. This change will require us to add "formal" `transp` values (just like `hcomp` is a constructor/value for HITs), for details see the special case of Id types in section 3.3 of the linked paper (note that `transp` is called `coe` as they are working in Cartesian cubical type theory, however the same idea works for De Morgan cubical type theory with `transp`/`hcomp`).

Reply:
@mortberg Shall we open the support of inductive families as an issue?

Reply:
Oh, don't worry, I'm aware we need them